Positioned errors

natefaubion opened the pull request:
This adds positions to the tokenizer. I've also gone through and reduced the pervasive backtracking, so the error locality is pretty good.

```
> print <$> parse """foo :=; SELECT"""
(Left (ParseError "Expected primary expression, got `;`" (Position { line: 1, column: 7 })))

> printQuery <$> parseQuery """IMPORT 42; SELECT * FROM foo"""
(Left (ParseError "Expected identifier, got integer literal" (Position { line: 1, column: 8 })))

> printModule <$> parseModule """IMPORT baz; CREATE FUNCTION foo"""
(Left (ParseError "Expected `(`, got EOF" (Position { line: 1, column: 29 })))
```

Needs a good 🔍 👀

natefaubion committed:
Positioned errors

cryogenian commented on src/SqlSquared/Parser.purs:
```diff
@@ -60,314 +88,227 @@ parseModule = tokenize >=> flip P.runParser (go <* eof)
+
+letExpr ∷ ∀ m t. SqlParser' m t
+letExpr = do
+  i ← PC.try (ident <* operator ":=")
```
Ooh! :+1:

cryogenian commented on src/SqlSquared/Parser.purs:
```diff
@@ -382,74 +323,62 @@ cases = PC.try do
   name ← ident
   operator "("
-  args ← PC.sepBy (operator ":" *> ident) $ operator ","
+  args ← PC.sepBy (operator ":" *> (ident <|> anyKeyword)) $ operator ","
```
Ah! Like "`foo`" is not ident, but `:in` is `operator ":" *> anyKeyword`! Nice catch

cryogenian commented on src/SqlSquared/Parser.purs:
```diff
@@ -459,157 +388,111 @@ functionDecl parseExpr = PC.try do
-  operator ":"
+pair ∷ ∀ m t. SqlParser m t (t × t)
+pair = do
+  l ← PC.try $ expr <* operator ":"
```
`pair` is used in `mapLiteral`, is there real reason for `PC.try` here? I.e. `{foo;bar}` doesn't make sense in next thrown error. Or it's for token position error?

natefaubion replied:
No, you are right. I was just going through things one-by-one locally 😆. I'll probably move this under the `mapLiteral` parser, then.

cryogenian commented on src/SqlSquared/Parser.purs:
```diff
@@ -676,63 +543,33 @@ tableRelation = PC.try do
+variRelation ∷ ∀ m t. SqlParser m t (Sig.Relation t)
+variRelation = do
   operator ":"
   vari ← ident
```
`create function foo(:in, :range)` is valid expression if we look at parser. Isn't `select * from :in` valid expression too? Maybe tokenizer should extract varis, what do you think?

natefaubion replied:
Do you mean `create function` is an expression in the scala parser or in this one? It shouldn't be in this one at least because they are only allowed in `SqlQuery` and `SqlModule`. I don't know about tokenizing varis. There are actually two problems:

* Currently, `:    foo` is a valid vari, which I think probably shouldn't be possible.
* If we tokenize vari, what do we do about `{ foo:bar }`. That's a valid map but would be tokens `{, foo, :bar, }`.

I don't think the backend parser tokenizes at all. Is there a reason that we don't just parse directly from the source string? I don't think we are getting anything interesting from our tokenizer.

cryogenian commented on src/SqlSquared/Parser.purs:
```diff
@@ -60,314 +88,227 @@ parseModule = tokenize >=> flip P.runParser (go <* eof)
+  fieldDeref = do
+    operator "."
+    k ← ident
+    pure \e → embed $ Sig.Binop
```
`pure $ _BINOP' Sig.FieldDeref e $ Sig.Ident k`?

cryogenian commented on src/SqlSquared/Parser.purs:
```diff
@@ -60,314 +88,227 @@ parseModule = tokenize >=> flip P.runParser (go <* eof)
+    operator "{"
+    rhs ← expr
+    operator "}"
+    pure \e → embed $ Sig.Binop
```
Same here. That's fine, but `_BINOP'` is much more readable

natefaubion commented on src/SqlSquared/Parser.purs:
```diff
@@ -60,314 +88,227 @@ parseModule = tokenize >=> flip P.runParser (go <* eof)
+  , keyword "union" $> Sig.Union
+  , keyword "union" *> keyword "all" $> Sig.UnionAll
+  , keyword "intersect" $> Sig.Intersect
+  , keyword "intersect" *> keyword "all" $> Sig.IntersectAll
```
Just realized this is wrong too, it will only ever parse "intersect", and never "intersect all".

natefaubion committed:
Merge remote-tracking branch 'origin' into positioned-try

jdegoes commented:
You should run this parser against all the queries in the regression tests in the quasar project. That will catch some more issues and give us more confidence.

natefaubion committed:
More cleanup, use constructors where available

natefaubion committed:
Add more tests. Fix bugs. Add error printer

natefaubion committed:
Add quasar regression test examples, bugfixes

natefaubion commented:
I've added all the quasar test queries, and they now parse. I've disabled the `Gen` tests for the Tokenizer, because they are broken. The generators build invalid SQL which will fail to tokenize. The only reason they were passing before was because the tokenizer always succeeded. It was missing an `eof` check, and so would always succeed with whatever it was able to consume, discarding the rest. I don't know what value the generators give us since we have a suite of real-world examples now, so I'm inclined to remove them rather than fix them. What do you think?

Also, I have changed the representation of `TableRelation` to allow unsandboxed relative paths. Many of the example failed because of this.

garyb commented:
> I don't know what value the generators give us since we have a suite of real-world examples now, so I'm inclined to remove them rather than fix them. What do you think?

I agree, sounds good. Unless we have some way of only generating valid SQL they don't seem very useful.

natefaubion committed:
Fix remaining bugs, more robust tests

natefaubion commented:
This is ready to go. I've made the parsing tests more robust by also printing and reparsing them, and making sure they parse to an equivalent AST. There were a lot of printer bugs, partly due to purescript-ejson being broken.

garyb reviewed (changes requested):
As per Slack conversation, need to escape keywords in `escapeIdent` also.

natefaubion committed:
Escape identifiers that match keywords

natefaubion commented:
Done
